File: StringObject.java
import java.awt.Color;
import java.awt.Font;
import java.awt.FontMetrics;
import java.awt.Graphics;

class StringObject extends DrawObject {
   private Graphics currentGra = null;
   private Font font;
   private String str;
   private Color color;
   private int x;
   private int y;
   boolean isUnderLine = false;
   private int strWidth;
   private int strHeight;
   private int align_ = 0;
   public static final int CENTER = 0;
   public static final int LEFT = 1;

   void draw(Graphics var1, DrawEnv var2) {
      if (this.str != null) {
         var1.setColor(this.color);
         var1.setFont(this.font);
         if (var1 != this.currentGra || this.strWidth < 0) {
            this.currentGra = var1;
            this.setStrSize();
         }

         int var3 = this.x;
         if (this.align_ == 0) {
            var3 = this.x - this.strWidth / 2;
         }

         var1.drawString(this.str, var3, this.y + this.strHeight / 2);
         if (this.isUnderLine) {
            var1.drawLine(var3, this.y + this.strHeight / 2 + 1, var3 + this.strWidth, this.y + this.strHeight / 2 + 1);
         }

      }
   }

   StringObject(Font var1, Color var2, String var3, int var4, int var5) {
      this.font = var1;
      this.str = var3;
      this.color = var2;
      this.x = var4;
      this.y = var5;
      this.strWidth = -1;
   }

   void setColor(Color var1) {
      this.color = var1;
   }

   void setAlign(int var1) {
      this.align_ = var1;
   }

   int getAlign() {
      return this.align_;
   }

   boolean hitTest(int var1, int var2) {
      if (this.currentGra == null) {
         return false;
      } else {
         return this.x - this.strWidth / 2 < var1 && var1 < this.x + this.strWidth / 2 && this.y - this.strHeight / 2 < var2 && var2 < this.y + this.strHeight / 2;
      }
   }

   private void setStrSize() {
      FontMetrics var1 = this.currentGra.getFontMetrics();
      this.strWidth = var1.stringWidth(this.str);
      this.strHeight = var1.getHeight();
   }

   void setText(String var1) {
      if (var1 != this.str) {
         if (var1 == null || !var1.equals(this.str)) {
            this.str = var1;
            this.strWidth = -1;
         }
      }
   }
}

File: original/js/stringObject.js
export const ALIGN_CENTER = 0;
export const ALIGN_LEFT = 1;

export class StringObject {
  constructor(font, color, str, x, y) {
    this.font = font;
    this.color = color;
    this.str = str;
    this.x = x;
    this.y = y;
    this.isUnderLine = false;
    this.align = ALIGN_CENTER;
    this._strWidth = -1;
    this._strHeight = -1;
    this._fontSize = parseInt(font) || 12;
  }

  _measure(ctx) {
    ctx.font = this.font;
    const m = ctx.measureText(this.str);
    this._strWidth = m.width;
    // Match Java's FontMetrics.getHeight() = ascent + descent
    this._strHeight = (m.fontBoundingBoxAscent || 0) + (m.fontBoundingBoxDescent || 0)
                      || this._fontSize;
  }

  draw(ctx) {
    if (this.str === null) return;

    if (this._strWidth < 0) {
      this._measure(ctx);
    }

    let drawX = this.x;
    if (this.align === ALIGN_CENTER) {
      drawX = this.x - this._strWidth / 2;
    }

    // Match Java: drawString(str, x, this.y + strHeight/2) — baseline positioning
    const baselineY = this.y + this._strHeight / 2;

    ctx.fillStyle = this.color;
    ctx.font = this.font;
    ctx.textBaseline = 'alphabetic';
    ctx.fillText(this.str, drawX, baselineY);

    if (this.isUnderLine) {
      const lineY = Math.round(baselineY + 1);
      ctx.fillStyle = this.color;
      ctx.fillRect(Math.round(drawX), lineY, Math.ceil(this._strWidth), 1);
    }
  }

  hitTest(mx, my) {
    if (this._strWidth < 0) return false;
    return (this.x - this._strWidth / 2 < mx && mx < this.x + this._strWidth / 2 &&
            this.y - this._strHeight / 2 < my && my < this.y + this._strHeight / 2);
  }

  setText(str) {
    if (str !== this.str) {
      this.str = str;
      this._strWidth = -1;
    }
  }

  setColor(color) {
    this.color = color;
  }

  setAlign(align) {
    this.align = align;
  }
}
